Offline mode implementation

cardenaso11 opened the pull request:
<!-- Describe your change here -->

TODO:
- [x] Remove commit https://github.com/input-output-hk/hydra/pull/1118/commits/4350c87ad34eabd9e55f4a08626ca21c10ea9671 (@ch1bo)

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [ ] Documentation updated or not needed
* [ ] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

locallycompact commented:
Do you expect a review now or do you want to wait?

cardenaso11 commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -86,15 +95,19 @@ main = do
           ctx <- loadChainContext chainConfig party otherParties hydraScriptsTxId
-          wallet <- mkTinyWallet (contramap DirectChain tracer) chainConfig
-          withDirectChain (contramap DirectChain tracer) chainConfig ctx wallet chainStateHistory (putEvent . OnChainEvent) $ \chain -> do
+          let headId = HeadId "FIXME(Elaine): headId"
```
I wanted to ask, it seems like the Head IDs are deterministically generated from on-chain transaction data, but as far as I can tell, it doesn't seem like their value would matter for the offline mode. Is that assesment correct?

cardenaso11 commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -127,7 +140,27 @@ main = do
+  withCardanoLedgerOffline OfflineConfig{ledgerGenesisFile} protocolParams action = do
+    --TODO(Elaine): double check previous messy branch for any other places where we query node
+    genesisParameters <- readJsonFileThrow (parseJSON @(Ledger.ShelleyGenesis StandardCrypto)) ledgerGenesisFile
+    globals <- newGlobals $ fromShelleyGenesis genesisParameters
```
We're reading in genesis parameters here, and while I don't think they're a huge obstacle to obtain, I think they're a necessity because of the type of applyTxes. Is this a correct assessment? Is there any way to sidestep this requirement currently?

cardenaso11 commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -136,3 +169,39 @@ main = do
+--TODO(ELAINE): figure out a less strange way to do this
+-- | Taken from Cardano.Api.GenesisParameters, a private module in cardano-api
+
+fromShelleyGenesis :: Shelley.ShelleyGenesis Ledger.StandardCrypto -> GenesisParameters ShelleyEra
```
If we really do need genesis parameters, is there a less awkward way to get them than parsing a ShelleyGenesis and running a copied fromShelleyGenesis on it? It's a private function so I'm not sure if this is exposing implementation details

cardenaso11 committed:
remove redundant comment

cardenaso11 commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -164,6 +178,68 @@ mkTinyWallet tracer config = do
       Consensus.interpreterToEpochInfo interpreter
 
+withOfflineChain ::
+  Tracer IO DirectChainLog -> -- TODO(ELAINE): change type maybe ?
```
Is it okay to reuse DirectChainLog here, or should we have a different type to indicate this is offline mode?

cardenaso11 commented:
> Do you expect a review now or do you want to wait?

A review now I think would be appropriate. Thank you

cardenaso11 committed:
validate offlinemode 0 peers, tick according to genesis, allow using persistence when offline, cleanup

cardenaso11 committed:
allow configuring writing utxo state to file

cardenaso11 committed:
do not require globals file to run offline mode

cardenaso11 committed:
fix tests, offline mode is offline, many changes

this commit might later be split up into several
for readability

cardenaso11 committed:
refactor offline chain into Hydra.Chain.Offline, cleanup

cardenaso11 committed:
organize and cleanup more offlinemode

cardenaso11 committed:
commit small todos i forgot to push earlier

cardenaso11 committed:
clean up draftCommitTx

cardenaso11 committed:
fix remaining todos

cardenaso11 commented:
This needs new tests to ensure the offline mode functionality but I think should otherwise be ready for review

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -161,6 +230,7 @@ data RunOptions = RunOptions
   , persistenceDir :: FilePath
   , chainConfig :: ChainConfig
   , ledgerConfig :: LedgerConfig
+  , offlineConfig :: Maybe OfflineConfig
```
I see that `Hydra.Chain.Offline` should be a drop-in replacement for `Hydra.Chain.Direct`. That makes a lot of sense to me.

Concretely, the `hydra-node` exe seems to switch a lot on the actual `Options` provided and does things differently.

I think it would be worthwhile to model this already her using something like
```haskell
data ChainConfig
  = DirectChainConfig { ... }
  | OfflineChainConfig { ... }
```

along with appropriate `Parser ChainConfig` which switches on the mutually exclusive command line options which determine the mode of operation.

My hunch is, that modelling it that way, will allow the `main` to separate more cleanly between the two modes.

ch1bo commented on hydra-node/golden/RunOptions.faulty.json:
Please avoid committing test output files. Both, `faulty.json` and `reencoded.json`, are outputs of the golden test suite which (rightfully) need updating when command line options change.

ch1bo commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -86,15 +94,41 @@ main = do
+                loadGlobalsFromGenesis ledgerGenesisFile
+
+        withCardanoLedger pparams globals $ \ledger -> do
+          persistence <- createStateChangePersistence (persistenceDir <> "/state") (offlineOptionsNormalizedUtxoWriteBackFilePath =<< leftToMaybe onlineOrOfflineConfig)
```
What's the rationale of this piggy-backing onto the standard persistence of `StateChanged` events?

It seems like you are interested in storing the latest `utxo` at some specific location (in addition to the `state` file), but I would like to know the whole user story for this.

---

This mechanism seems quite orthogonal to the offline mode chain layer and I'd prefer having this in a dedicated PR (in which I would ask for the purpose of this the same way).

Depending in how the downstream usage of this file is, a more modular alternative could be a small utility subscribing to the API (websocket) and writing this file on every `SnapshotConfirmed` message.

Another alternative would be to make this "writeBack" mechanism be attached as a consumer of those `StateChanged` events [here](https://github.com/input-output-hk/hydra/blob/8a451dd0721b2f28050b57e527bcd376e9ca291e/hydra-node/src/Hydra/Node.hs#L188). This would support also an idea we had where the `Effects` of the business logic (e.g. sending output to clients on the API) be mere interpretations of the event stream.

Quantumplation replied:
I think the main concern is that in offline mode, we don't have any "Commit" events, so it's decidedly useless: the head opens with an empty UTXO, and so we can't spend any transactions;

So, we added the ability to specify the initial UTXO on the command line, but in a call we had with the team, there was concerns about *where to get this initial UTXO file in the first place*; as a convenience, we agreed to add something that saved the UTXO state out when the node shut down gracefully, so we could reuse that on the next run.

If it feels a bit messy / arbitrary, we can strip that back out and say that you need to construct the UTXO by hand, and/or fold this into our other ADR (a PR for which is coming soon!)

ch1bo replied:
I see. But why would you open it empty? Also wouldn't the normal persistence of internal state have the `hydra-node` start up in the `Open` state with just the same UTxO as on the last run?

In my mind the offline mode could work like the following:
* Always simulates `Init` and `Commit` observations based on some configured initial utxo file
* When no previous state is loaded, the head logic would interpret that as opening of the head
* Allow normal off-chain transaction processing, evolving the `Open` off-chain head state (which is automatically persisted)
* On later starts, the head would already be in an `Open` state as loaded from persistence and any (simulated) `Init` and `Commit` observations from the offline chain would be ignored

The initial UTxO file would be given by your outer gummiworm protocol I suppose.

ch1bo commented on hydra-node/src/Hydra/Chain/Offline/Handlers.hs:
```diff
@@ -0,0 +1,47 @@
+                      OnContestTx{snapshotNumber = number $ getSnapshot confirmedSnapshot}}
+                FanoutTx{} ->
+                  callback $ Observation { newChainState = cst, observedTx =
+                      OnFanoutTx{}}
```
Can an offline chain even open/close a head? I would have expected the offline chain to be in an `Open` state right away (constantly yielded `Observation` when starting it) and also ignore all commands to `CloseTx` it.

ch1bo commented on hydra-node/src/Hydra/Chain/Offline.hs:
```diff
@@ -0,0 +1,131 @@
+withOfflineChain tracer OfflineConfig{ledgerGenesisFile} globals@Ledger.Globals{systemStart} ownHeadId chainStateHistory callback action = do
+
+
+  localChainState <- newLocalChainState chainStateHistory
```
As you point out in a NOTE further up, we do not need to keep a chain state in the offline chain layer, as it is not needed to construct transactions (in fact there are no L1 transactions). So we can keep it `initialChainState = ChainStateAt{chainState = Idle, recordedAt = Nothing}`

ch1bo replied:
If you apply my suggestions, this becomes unused.

ch1bo commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -86,15 +94,41 @@ main = do
+            Nothing -> pure ()
+            Just (OfflineConfig {initialUTxOFile})-> do
+              initialUTxO :: UTxOType Tx <- readJsonFileThrow (parseJSON @(UTxOType Tx)) initialUTxOFile
+              initializeStateIfOffline chainStateHistory initialUTxO headId party contestationPeriod (putEvent . OnChainEvent)
```
I think this (yielding of constant observations to simulate a head opening) can be moved to `withOfflineChain` directly, which would simplify the main function (less casing).

ch1bo reviewed (commented):
Very nice work so far. As you already mentioned, tests are definitely needed for this and I left a few more remarks on what could be improved.

When you rebase onto the latest `master`, you can ensure your changes are formatted correctly (which is checked by the CI now) by simply calling `treefmt` from the nix shell.

cardenaso11 committed:
various pr feedback

cardenaso11 committed:
code standard

cardenaso11 committed:
dump worwk for the night

cardenaso11 committed:
temp commit

ghost commented:
@cardenaso11 How are things going? Is this PR ready for review, or perhaps another pairing session could be helpful?

cardenaso11 committed:
bugfixes, end to end offline mode tests

cardenaso11 commented:
> @cardenaso11 How are things going? Is this PR ready for review, or perhaps another pairing session could be helpful?

I think it is ready for review

ghost commented:
@cardenaso11 Could you rebase the branch on latest master and resolve conflicts?

cardenaso11 committed:
initial untested gummiworm-redacted mostly-cleaned-up offline mode implementation

ghost commented:
I noticed a few tests failing, would you need help on this @cardenaso11 ?

cardenaso11 commented:
@abailly-iohk should be fixed now, had some nix changes for dealing with arm64 macos I didn't realize I checked in that got mangled during rebase somehow

cardenaso11 committed:
remove re-checked-in golden test failures, add to gitignore since they shouldn't be added generally

cardenaso11 committed:
commit only test fixes

ch1bo commented:
Rebased this branch onto latest `master` and force-pushed.

The CI was not running since it had merge conflicts. It seems that a prior rebase already broke things into a non-compiling state.

I will commence in making it compile next.

Quantumplation commented:
@ch1bo Thanks for doing the rebase; We can take a pass at getting the branch green, the rebase was already a generous use of your time :sweat_smile:

That being said, if it's super quick and you've already done it, no worries.

ch1bo committed:
Fix compilation, several warnings remain

Only fixes compilation somewhat as the warnings need to be addressed for
making it work properly.

ch1bo committed:
Minimize diff to master

There were several things changed forth and back without ultimately
changing anything on this branch. This commit aims to reduce the diff to
master to make the actual change more visible.

ch1bo commented:
@Quantumplation @cardenaso11
 
> getting the branch green

I think there is some work to be done still. I have only fixed the compile errors, there are still a lot of warnings which will guide us in dealing with some things that we definitely take care of before we can hope to see the [offline mode end-to-end test green](https://github.com/input-output-hk/hydra/actions/runs/7198189068/job/19607057726?pr=1118#step:5:990).

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -281,6 +281,81 @@ withConfiguredHydraCluster tracer workDir nodeSocket firstNodeId allKeys hydraKe
+                { verbosity = Verbose "HydraNode"
+                , nodeId = NodeId $ show hydraNodeId
+                , host = "127.0.0.1"
+                , -- NOTE(Elaine): port 5000 is used on recent versions of macos
```
Hm, this is going to be a problem for us .. was this resulting in existing tests to fail on you (i.e. on `master`)? Could you please file an issue for that alone?

ch1bo commented:
Pushed https://github.com/input-output-hk/hydra/pull/1118/commits/4350c87ad34eabd9e55f4a08626ca21c10ea9671 to unblock the CI not failing on `hydra-node` golden files. It's currently unclear what change in this PR had the transaction be de-/serialized slightly differently. We are currently investigating this, but now the `hydra-cluster` tests should run on this branch and we can focus on that.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -112,7 +114,35 @@ withClusterTempDir name =
+            ]
+      (aliceCardanoVk, aliceCardanoSk) <- keysFor Alice
+      (bobCardanoVk, _) <- keysFor Bob
+      offlineConfig <- offlineConfigFor startingState tmpDir networkId
```
Should any arbitrary `UTxO` work to be loaded by the `offline` mode?

I think we should encode that (and make it visible) in this test by doing something like this inline:

```hs
initialUTxO <- generate $ do
  a <- genUTxOFor aliceCardanoVk
  b <- genUTxOFor bobCardanoVk
  pure $ a <> b
Aeson.encodeFile (tmpDir </> "utxo.json") initialUTxO
let offlineConfig = OfflineConfig
      { initialUTxOFile = tmpDir </> "utxo.json"
      , ledgerGenesisFile = Nothing
      }
``` 

Of course this will test some random UTxO each time (it's not a property test), but should already provide a better coverage than only a single ADA value (e.g. other assets)

Also, this is more clear in that the offline mode loads this file (one can read the test to know and maintain this, instead of all functions behind `offlineConfigFor`).

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -112,7 +114,35 @@ withClusterTempDir name =
+      offlineConfig <- offlineConfigFor startingState tmpDir networkId
+
+      initialUtxo <- Aeson.throwDecodeStrict @UTxO.UTxO =<< readFileBS (initialUTxOFile offlineConfig)
+      let Just (aliceSeedTxIn, aliceSeedTxOut) = UTxO.find (\(TxOut addr _ _ _) -> addr == mkVkAddress networkId aliceCardanoVk) initialUtxo
```
Could: use scenario values instead of loading/inferring them again here.

For example given my comment above, this could be simplified by just keeping `aliceInitialUTxO <- generate $ genUTxOFor aliceCardanoVk` around and there is also `UTxO.min` which will give you one (no filtering needed if we generate only for alice).

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -281,6 +281,81 @@ withConfiguredHydraCluster tracer workDir nodeSocket firstNodeId allKeys hydraKe
+                , hydraVerificationKeys = []
+                , hydraScriptsTxId = "9fdc525c20bc00d9dfa9d14904b65e01910c0dfe3bb39865523c1e20eaeb0903"
+                , persistenceDir = workDir </> "state-" <> show hydraNodeId
+                , chainConfig = defaultChainConfig
```
None of `hydraScriptsTxId` or `chainConfig` (the direct chain one) should be needed here.

In fact, the end-to-end test is currently failing with:

```
hydra-node: user error (FileDoesNotExistError "cardano.sk")
```

Suggesting that we are currently trying to load the `--cardano-signing-key` (via the default chain config). It might have happened only since our rebase, but this confirms to me that we would want to make this situation _impossible_ by using the type system.

---

The error above is only visible after doing some plumbing modifications on this function to actually inherit the stderr like we had done on the `withHydraNode` in the mean-time. Which suggests that it's probably not a good idea to replicate this `withHydraNode` into the `withOfflineHydraNode` in the first place. We would like to pass in the chain config switching the modes instead.

ch1bo reviewed (commented):
Left some comments where I think the test is currently failing and if wished I could fix it.

Also, I would like to make the end-to-end scenario cover all your expectations and see it as a contract between us (so we don't break your application when we continue working on the code base). Maybe there is more that should stay working after this is merged.

For example:
* do we expect the offline mode head to be reporting `HeadIsOpen` on every start?
* do we expect the offline mode to reject any `Close` requests? 
* how does the `hydra-node` in offline mode behave on restarts? (currently it will just pick off where it was.. I think)

cardenaso11 committed:
WIP: get tests passing, with debug code left in

cardenaso11 committed:
demonstrate offline mode running with arbitrary participant OnChainId

cardenaso11 committed:
clean up most stuff, get rid of debug code & logs

cardenaso11 committed:
Revert "demonstrate offline mode running with arbitrary participant OnChainId"

This reverts commit 84adba8663574c01f41e1fd2e4d806c2b9184acb.

cardenaso11 commented:
Have some more changes on this I'll be pushing up shortly, addressing some of the other feedback besides just getting tests passing

cardenaso11 commented:
One important thing is that it wasn't immediately clear to me whether the new logic-level participants validation should be empty in offline mode, or if it should be a singleton list of a hard-coded value. The tests pass with either solution, implemented in 84adba8 and reverted in b3562de.

I think the best solution that doesn't kludge the meaning of OnChainId is an empty list, that is the solution that's in here right now, but the Hydra team is welcome to merge without the revert patch. Right now I don't think there would be anything expecting an OnChainId participant to have a particular form (in offline mode), but I could imagine that that might be something used at some point, somehow

cardenaso11 committed:
wip: temporary commit for demoing offline mode. Good for demo, do not merge as is

cardenaso11 committed:
Revert "wip: temporary commit for demoing offline mode. Good for demo, do not merge as is"

This reverts commit 6599c0a9e3437361395ea8dce8d167af9f32b852.

ch1bo commented on .gitignore:
```diff
@@ -24,6 +24,8 @@ result*
 test-results.xml
 hspec-results.md
+hydra-node/golden/RunOptions.faulty.json
+hydra-node/golden/RunOptions.faulty.reencoded.json
```
Must: remove this.

ch1bo commented on cabal.project:
```diff
@@ -35,7 +35,7 @@ package *
 -- Warnings as errors for local packages
 program-options
-  ghc-options: -Werror
+  ghc-options: -Wwarn
```
Must: not change the project config. See https://github.com/input-output-hk/hydra/blob/master/CONTRIBUTING.md#building--testing for how to have a project.local which does not fail on warnings.

cardenaso11 committed:
wip: separate top-level command and run(offline)options type. needs to have conflicts minimized, warnings and debug code removed. this'll get rebased

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Util.hs:
```diff
@@ -59,6 +78,62 @@ createAndSaveSigningKey path = do
+  let destinationPath = targetDir </> "utxo.json"
+  writeFileBS destinationPath . toStrict . Aeson.encode $ utxo
+
+  -- Aeson.throwDecodeStrict =<< readFileBS (targetDir </> "utxo.json")
```
```suggestion
  -- Aeson.throwDecodeStrict =<< readFileBS (targetDir </> "utxo.json")
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Util.hs:
```diff
@@ -1,30 +1,49 @@
 import Test.QuickCheck (generate)
 
+-- import CardanoClient (buildAddress)
+
```
```suggestion
-- import CardanoClient (buildAddress)

```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Util.hs:
```diff
@@ -73,7 +148,7 @@ chainConfigFor me targetDir nodeSocket them cp = do
       , cardanoSigningKey = skTarget me
       , cardanoVerificationKeys = [vkTarget himOrHer | himOrHer <- them]
-      , contestationPeriod = cp
+      , contestationPeriod = cp :: ContestationPeriod
```
```suggestion
      , contestationPeriod
```

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -349,8 +423,11 @@ withHydraNode' chainConfig workDir hydraNodeId hydraSKey hydraVKeys allNodeIds h
+                  chainConfig
                 , ledgerConfig
+                , -- , offlineConfig = leftToMaybe chainConfig
+                  offlineConfig = Nothing
```
```suggestion
                , offlineConfig = Nothing
```

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -349,8 +423,11 @@ withHydraNode' chainConfig workDir hydraNodeId hydraSKey hydraVKeys allNodeIds h
                 , persistenceDir = workDir </> "state-" <> show hydraNodeId
-                , chainConfig
+                , -- , chainConfig = fromRight defaultChainConfig chainConfig
+                  chainConfig
```
```suggestion
                , chainConfig
```

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -112,7 +114,35 @@ withClusterTempDir name =
+        waitMatch 10 node $ \v -> do
+          guard $ v ^? key "tag" == Just "SnapshotConfirmed"
+
+        pure ()
```
```suggestion
```

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -181,6 +184,7 @@ executable hydra-node
     , hydra-cardano-api
     , hydra-node
     , hydra-prelude
+    , base
```
```suggestion
```
Should not be needed as our prelude contains most of `base`.

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -147,6 +147,8 @@ data PostTxError tx
   | -- | We can only draft commit tx for the user when in Initializing state
     FailedToDraftTxNotInitializing
+  | -- | We cannot draft a commit tx in offline mode
+    FailedToDraftTxOffline
```
```suggestion
```

Unused code

ch1bo commented on hydra-node/src/Hydra/Chain/Offline/Persistence.hs:
```diff
@@ -0,0 +1,88 @@
+              { headId = ownHeadId
+              , headParameters = HeadParameters{parties = [ownParty], contestationPeriod}
+              , headSeed = UnsafeHeadSeed "OfflineHeadSeed_" -- FIXME(Elaine): might want to generate?
+              , participants = []-- error "Participants unimplemented!" -- FIXME(Elaine): might want to make arbitrary / garbage 28 byte value
```
I think having an empty list of participants is fine, you can remove those comments.

ch1bo commented on hydra-node/src/Hydra/Chain/Offline/Persistence.hs:
```diff
@@ -0,0 +1,88 @@
+              , committed = initialUTxO
+              , headId = ownHeadId
+              }
+        }
```
```suggestion
        }
    callback $
      Observation
        { newChainState = initialChainState
        , observedTx = OnCollectComTx{headId = ownHeadId}
        }

```

If you add a `OnCollectComTx` here, you can remove the whole `mkFakeL1Chain` and make it just do `pure ()` on `postChainTx`. I'll suggest a comment over there as well.

ch1bo commented on hydra-node/src/Hydra/Chain/Offline.hs:
```diff
@@ -0,0 +1,138 @@
+  initializeStateIfOffline chainStateHistory initialUTxO ownHeadId party contestationPeriod callback
+
+  localChainState <- newLocalChainState chainStateHistory
+  let chainHandle = mkFakeL1Chain contestationPeriod localChainState tracer ownHeadId callback
```
```suggestion
  let chainHandle =
        Chain
          { submitTx = const $ pure ()
          , draftCommitTx = \_ _ -> pure $ Left FailedToDraftTxNotInitializing
          , postTx = const $ pure ()
          }
```

If we add `OnCollectComTx` as commented in `initializeStateIfOffline`, we do not need to have a working `Chain` handle at all. We can just do nothing (or fail) if a user tries to `Close`.

ch1bo commented on hydra-node/src/Hydra/Chain/Offline.hs:
```diff
@@ -0,0 +1,138 @@
+              Tick
+                { chainTime = timeToSleepUntil
+                , chainSlot = ChainSlot . fromIntegral @Word64 @Natural $ unSlotNo upcomingSlot
+                }
```
I think this can be greatly simplified as there will be no epochs and no differring slot lengths (you use `neverForksSummary`). That is, the current slot can be just computed by lookging at `currentTime` and `startTime` from the genesis shelley and devide that by the `slotLength`. No need for the epoch summary interpreters. For the ticking itself you could do something like

```hs
let tickForeverAction = tick slotLengthNominalDiffTime callback startSlot
```
with

```hs
tick :: (MonadDelay m, MonadTime m) => NominalDiffTime -> (ChainEvent tx -> m ()) -> Natural -> m ()
tick slotLengthNominalDiffTime callback start = do
  go start
 where
  go !n = do
    threadDelay $ realToFrac slotLengthNominalDiffTime
    chainTime <- getCurrentTime
    callback $
      Tick
        { chainTime
        , chainSlot = ChainSlot n
        }
    go $ n + 1
```

ch1bo commented on hydra-node/src/Hydra/Chain/Offline/Persistence.hs:
```diff
@@ -0,0 +1,88 @@
+            SnapshotConfirmed{snapshot = Snapshot{utxo}} ->
+              writeBinaryFileDurableAtomic utxoFilePath . toStrict $ Aeson.encode utxo
+            _ -> pure ()
+      }
```
```suggestion
```

Unused code

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -65,9 +73,17 @@ initEnvironment options = do
+    
   otherVerificationKeys <- mapM (readFileTextEnvelopeThrow (AsVerificationKey AsPaymentKey)) cardanoVerificationKeys
-  let participants = verificationKeyToOnChainId <$> (getVerificationKey ownSigningKey : otherVerificationKeys)
+  -- let participants = []
```
```suggestion
```

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -65,9 +73,17 @@ initEnvironment options = do
+      -- online mode
+    Nothing -> readFileTextEnvelopeThrow (AsSigningKey AsPaymentKey) cardanoSigningKey
+      -- offline mode
+    Just _ -> pure $ Hydra.Prelude.error "Shouldn't be using cardanoSigningKey in offline mode!"
```
```suggestion
    Just _ -> die "Shouldn't be using cardanoSigningKey in offline mode!"
```

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -282,3 +299,56 @@ loadState tracer persistence defaultChainState = do
+          Shelley.Lovelace
+            (fromIntegral sgMaxLovelaceSupply)
+      , protocolInitialUpdateableProtocolParameters = Shelley.sgProtocolParams sg
+      }
```
Should: Please move the genesis file loading into the `Hydra.Chain.Offline` module. It's only used there.

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -80,15 +90,37 @@ run opts = do
+        let withChain cont = case onlineOrOfflineConfig of
+              Left offlineConfig' ->
+                let headId = UnsafeHeadId "HeadId"
+                 in withOfflineChain (contramap DirectChain tracer) offlineConfig' globals headId party contestationPeriod chainStateHistory (putEvent . OnChainEvent) cont
```
I would suggest to hard-code the `headId` and `contestationPeriod` in the offline chain. This will have you pass less arguments around.

Also if you follow my other suggestions, `globals` will be unused.

ch1bo reviewed (commented):
There is a lot of improvement potential, but for this to be merged please do at least:
* Changelog entry
* Don't change cabal.project to not error on warnings -> must remove all warnings
* Remove commented and unused code
* Resolve or explain any TODO in the PR description

Also, I am not sure if this works correctly if the `hydra-node` is restarted after processing a transaction? Maybe an additional  or extended EndToEnd test would ensure this.

cardenaso11 committed:
pr feedback

cardenaso11 committed:
revert back to old port: see #1218 for macOS port bug

cardenaso11 committed:
more pr feedback

cardenaso11 committed:
more feedback, undo error->die feedback due to introduced test failure

rrruko committed:
add to changelog

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
randomize utxo in offline test

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
remove dead code

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
remove TODOs

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
Hydra.Options: import toArgs from Online

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
resolve warnings

Co-authored-by: card <cardenaso11@users.noreply.github.com>

rrruko committed:
update log schema

rrruko committed:
fix test failures

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -266,19 +286,3 @@ createNodeState initialState = do
-      chainStateHistory = recoverChainStateHistory defaultChainState events
-  pure (headState, chainStateHistory)
- where
-  initialState = Idle IdleState{chainState = defaultChainState}
```
Must: not move this out of the `Node` module.

Why did we move this into the `Hydra.Chain.Offline` module now? This is completely orthogonal change and makes this PR longer than it should be.

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -66,6 +80,68 @@ explain = \case
+      pparams <- case toLedgerPParams ShelleyBasedEraBabbage protocolParams of
+        Left err -> throwIO (ConfigurationException err)
+        Right bpparams -> pure bpparams
+      -- let DirectChainConfig{networkId, nodeSocket} = chainConfig
```
Must: remove commented code

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -66,6 +80,68 @@ explain = \case
+
+  withCardanoLedger protocolParams globals action =
+    let ledgerEnv = newLedgerEnv protocolParams
+     in action (Ledger.cardanoLedger globals ledgerEnv)
```
Should: not separate the whole `run` semantics just to replace the `withChain` with the offline-mode.

This leads to unnecessary duplication and makes maintainability of `hydra-node` harder as it could be.

This does not prevent merging, but expect major refactor back to have a single `run` again.

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -115,12 +199,48 @@ run opts = do
+          Shelley.Lovelace
+            (fromIntegral sgMaxLovelaceSupply)
+      , protocolInitialUpdateableProtocolParameters = Shelley.sgProtocolParams sg
+      }
```
Must: remove duplicate code. 

This function does exist in `Hydra.Chain.Offline` now (good!), but seemingly this is a remainder which should be unused now.

ch1bo commented on hydra-node/src/Hydra/Options/Offline.hs:
```diff
@@ -0,0 +1,207 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
Should: remove unused `no-orphans` ghc option

I do not see why this would be needed in this module.

ch1bo commented on hydra-node/src/Hydra/Options/Offline.hs:
```diff
@@ -0,0 +1,207 @@
+
+module Hydra.Options.Offline (
+  module Hydra.Options.Offline,
+) where
```
This explicit export list is pointless. We are usually fine by exporting everything so this could be removed.

ch1bo commented on hydra-node/src/Hydra/Options/Online.hs:
```diff
@@ -0,0 +1,463 @@
+module Hydra.Options.Online (
+  module Hydra.Options.Online
+ ) where
```
```suggestion
module Hydra.Options.Online where
```

ch1bo reviewed (changes requested):
While recent commits made the tests pass, there are still **formatting** changes needed, some dead code .. and a bit puzzling to me: moved `loadState` function?

The way the `Hydra.Options` are separated now into two modules and the duplicated `run` for online/offline makes this not a very desirable, but otherwise mergeable PR. Do expect another PR refactoring things like these back into other places just after (no semantic changes of course).

Please address my "Must" comments.

ch1bo committed:
Fix formatting

rrruko committed:
address some feedback

rrruko committed:
deduplicate fromShelleyGenesis

rrruko committed:
move back loadState

rrruko committed:
organize imports

ch1bo reviewed (approved):
All must issues are resolved now and I can approve merging this.

I will follow-up with a PR that should significantly simplify the `Hydra.Options` handling and downstream code duplication in `Hydra.Node.Run` and `HydraNode` modules.

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -281,6 +283,73 @@ withConfiguredHydraCluster tracer workDir nodeSocket firstNodeId allKeys hydraKe
+ where
+  logFilePath = workDir </> "logs" </> "hydra-node-" <> show hydraNodeId <.> "log"
+
+withOfflineHydraNode' ::
```
I'll cover this in a refactor PR.
